modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DBBasedExpiryMessageDeletionTask.java

```diff
@@ -0,0 +1,169 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wso2.andes.kernel;
+
+import com.google.common.util.concurrent.SettableFuture;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.andes.configuration.AndesConfigurationManager;
+import org.wso2.andes.configuration.enums.AndesConfiguration;
+import org.wso2.andes.kernel.slot.SlotManagerClusterMode;
+import org.wso2.andes.kernel.slot.SlotManagerStandalone;
+import org.wso2.andes.store.FailureObservingStoreManager;
+import org.wso2.andes.store.HealthAwareStore;
+import org.wso2.andes.store.StoreHealthListener;
+
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * This task is responsible to delete the expired messages from the database which were not currently
+ * allocated to any slots. If we delete any messages that are assigned to a slot delivery worker, the
+ * content may got deleted when the content is asked to retrieve. This can mess up the message delivery
+ * flow. Because of that the messages not allocated to a slot are considered as safe to delete
+ */
+public class DBBasedExpiryMessageDeletionTask implements Runnable, StoreHealthListener {
+
+    private static Log log = LogFactory.getLog(SetBasedExpiryMessageDeletionTask.class);
+
+    /**
+     * Indicates and provides a barrier if messages stores become offline.
+     * marked as volatile since this value could be set from a different thread
+     */
+    private volatile SettableFuture<Boolean> messageStoresUnavailable;
+
+    /**
+     * Used to perform database operations on the context store.
+     */
+    private AndesContextStore andesContextStore = AndesContext.getInstance().getAndesContextStore();
+
+    /**
+     *  The slot gap need to be maintained as not reachable for deletion task as they may be allocated
+     *  to a slot delivery worker in the near future
+     */
+    private final int safetySlotCount;
+
+
+    public DBBasedExpiryMessageDeletionTask() {
+
+        this.messageStoresUnavailable = null;
+        // Register AndesRecoveryTask class as a StoreHealthListener
+        FailureObservingStoreManager.registerStoreHealthListener(this);
+
+        safetySlotCount = AndesConfigurationManager.readValue
+                (AndesConfiguration.PERFORMANCE_TUNING_SAFE_DELETE_REGION_SLOT_COUNT);
+
+    }
+
+    @Override
+    public void run() {
+
+        boolean isClusterEnabled = AndesContext.getInstance().isClusteringEnabled();
+
+        try {
+            /**
+             * This logic belongs to an MB run in stand alone mode / the coordinator node run in cluster mode
+             */
+            if (!isClusterEnabled || (isClusterEnabled && AndesContext.getInstance().getClusterAgent().isCoordinator())) {
```

Comment:
We can do this calculation at the initialization of the class and store it to a field like 'isCoordinatorInCluster', and then have something like 'if (isCoordinatorInCluster)' here